keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -1235,374 +1267,326 @@ const (
+pod already got deleted because such objects are orphaned
+and will be removed soon.
+
+```
+// PodSchedulingSpec describes where resources for the Pod are needed.
+type PodSchedulingSpec struct {
+	// SelectedNode is the node for which allocation of ResourceClaims that
+	// are referenced by the Pod and that use "WaitForFirstConsumer"
+	// allocation is to be attempted.
 	SelectedNode string
 
-	// When allocation is delayed, and the scheduler needs to
-	// decide on which node a Pod should run, it will
-	// ask the driver(s) on which nodes the resource might be
-	// made available. To trigger that check, the scheduler
-	// provides the names of nodes which might be suitable
-	// for the Pod. Will be updated periodically until
-	// all resources are allocated.
-	//
-	// The ResourceClass.SuiteableNodes node selector can be
-	// used to filter out nodes based on labels. This prevents
-	// adding nodes here that the driver then would need to
-	// reject through UnsuitableNodes.
+	// PotentialNodes lists nodes where the Pod might be able to run.
 	//
-	// The size of this field is limited to 256 (=
-	// [PodSchedulingNodeListMaxSize]).  This is large enough for many
-	// clusters. Larger clusters may need more attempts to find a node that
-	// suits all pending resources.
+	// The size of this field is limited to 128. This is large enough for
+	// many clusters. Larger clusters may need more attempts to find a node
+	// that suits all pending resources. This may get increased in the
+	// future, but not reduced.
+	// +optional
 	PotentialNodes []string
 }
+```
+
+When allocation is delayed, the scheduler must set
+the `SelectedNode` for which it wants the resource(s) to be allocated
+before the driver(s) start with allocation.
+
+The driver must ensure that the allocated resource
+is available on this node or update ResourceSchedulingStatus.UnsuitableNodes
+to indicate where allocation might succeed.
+
+When allocation succeeds, drivers should immediately add
+the pod to the ResourceClaimStatus.ReservedFor field
+together with setting ResourceClaimStatus.Allocated. This
+optimization may save scheduling attempts and roundtrips
+through the API server because the scheduler does not
+need to reserve the claim for the pod itself.
+
+The selected node may change over time, for example
+when the initial choice turns out to be unsuitable
+after all. Drivers must not reallocate for a different
+node when they see such a change because it would
+lead to race conditions. Instead, the scheduler
+will trigger deallocation of specific claims as
+needed through the ResourceClaimStatus.DeallocationRequested
+field.
+
+When allocation is delayed, and the scheduler needs to
+decide on which node a Pod should run, it will
+ask the driver(s) on which nodes the resource might be
+made available. To trigger that check, the scheduler
+provides the names of nodes which might be suitable
+for the Pod. Will be updated periodically until
+all resources are allocated.
```

Comment:
I think this paragraph would flow better if added as part of the paragraph above that reads:

> When allocation is delayed, the scheduler must set
> the `SelectedNode` for which it wants the resource(s) to be allocated
> before the driver(s) start with allocation.

As part of that, I'd drop the redundant "When allocation is delayed" and just start with:
> The scheduler also needs to decide on which node a Pod should run and will ...

Otherwise you start talking about UnsuitableNodes before introducing how the driver is given the opportunity to set this field.

Reply:
Makes sense, changed.